File: day-5/soln.py
def to_binary(ticket):
    ticket = ticket.replace('F', '0')
    ticket = ticket.replace('B', '1')
    ticket = ticket.replace('L', '0')
    ticket = ticket.replace('R', '1')

    return int(ticket, 2)


def part1(ipt):
    return max([to_binary(ticket) for ticket in ipt])


def part2(ipt):
    all_seats = range(0, 0b10000000000)

    for t in ipt:
        as_binary = to_binary(t)

        all_seats = [x for x in all_seats if x != as_binary]

    for i, s in enumerate(all_seats):
        if i == 0 or i == len(all_seats) - 1:
            continue

        if s != all_seats[i+1] - 1 and s != all_seats[i-1] + 1:
            return s

    return list(all_seats)


if __name__ == '__main__':
    with open('input.txt') as fp:
        print(part1(fp))
        fp.seek(0)
        print(part2(fp))

File: day-5/soln.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

var binMap = map[string]string{
	"F": "0",
	"B": "1",
	"L": "0",
	"R": "1",
}

func toBinary(ticket string) (asBinary uint64) {
	for old, new := range binMap {
		ticket = strings.Replace(ticket, old, new, -1)
	}

	asBinary, err := strconv.ParseUint(ticket, 2, 64)
	if err != nil {
		log.Fatal(err)
	}

	return
}

func part1(ipt []string) (max uint64) {
	for _, ticket := range ipt {
		asBinary := toBinary(ticket)
		if asBinary > max {
			max = asBinary
		}
	}

	return
}

func part2(ipt []string) uint64 {
	for i := 1; i < len(ipt) - 1; i++ {
		prev := toBinary(ipt[i-1])
		curr := toBinary(ipt[i])
		next := toBinary(ipt[i+1])

		if prev != curr - 1 && next != curr + 1 {
			return curr
		}
	}

	return 0
}

func main() {
	var ipt []string

	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		ipt = append(ipt, scanner.Text())
	}

	fmt.Println(part1(ipt))
	fmt.Println(part2(ipt))
}